src/Transformer/AnnotationTransformer.php

```diff
@@ -0,0 +1,215 @@
+<?php declare(strict_types=1);
+
+namespace SilverStripe\Config\Transformer;
+
+use InvalidArgumentException;
+use ReflectionClass;
+use SilverStripe\Config\Collections\MutableConfigCollectionInterface;
+use SilverStripe\Config\Transformer\AnnotationTransformer\AnnotationDefinitionInterface;
+
+class AnnotationTransformer implements TransformerInterface
+{
+    /**
+     * Set a callable that can be used to resolve the classes to collect annotations from
+     *
+     * @var callable
+     */
+    protected $classResolver;
+
+    /**
+     * An array of @see AnnotationDefinitionInterface that indicate what annotations are collected from doc blocks and
+     * how those annotations are converted to config
+     *
+     * @var AnnotationDefinitionInterface[]
+     */
+    protected $annotationDefinitions;
+
+    /**
+     * The list of resolved class names once the resolver has been called
+     *
+     * @var array|null
+     */
+    protected $classes;
+
+    /**
+     * AnnotationTransformer constructor.
+     * @param callable $classResolver
+     * @param array $annotationDefinitions
```

Comment:
You could annotate `AnnotationDefinitionInterface[]` here to make it clearer what is expected. Also worth adding a `@throws` for the exception if one of the entries isn't one of these interfaces